design/ceph/osd-migration.md

```diff
@@ -0,0 +1,139 @@
+# OSD Backend Store Migration
+
+## Summary
+
+- Configure OSD backend store via `cephCluster` CR so that new OSDs can use it.
+- Existing OSDs should be able to migrate to use a new backend store configured via `cephCluster` CR. The migration process should include destroying the existing OSDs one by one, wiping the disk and then creating a new OSD on that disk.
+
+## Motivation:
+
+- Ceph uses BlueStore as a special-purpose storage back end designed specifically for managing data on disk for Ceph OSD workloads. Ceph team is working on improving this backend storage that will help in significantly improving the performance, especially in disaster recovery scenarios. This new OSD backend is known as `bluestore-rdr`. In the future, other backends will also need to be supported, such as `seastore`.
+- To support any new and improved storage backend, the existing OSD needs to be destroyed, disks should be wiped out and new OSD should be created using the latest storage backend.
+
+## Goals
+
+- Newly provisioned clusters should be able to use the specific storage backend while creating OSDs.
+- Existing clusters should be able to migrate the OSDs to use a specific storage backend, without downtime and without risking data loss.
+
+## Non-Goals/Deferred Features:
+
+Migration of OSDs with following configurations is deferred for now and will be taken up in future:
+- OSDs on mixed media setup where metadata (RocksDB+WAL) is placed on higher throughput storage media and application data on lower throughput media.
+- Setups with multiple OSDs per disk.
+
+## Proposal
+
+### API Changes
+- Add `spec.storage.store` in the ceph cluster yaml.
+
+  ```yaml
+  storage:
+    store:
+      type: bluestore-rdr
+      updateStore: yes-really-update-store
+  ```
+
+  - `type`: The type of backend store to be used for OSDs. For example, bluestore, bluestore-rdr, etc. The default store type will be `bluestore`
+  - `updateStore`: Allows the user to migrate the existing OSDs to use any new storage backend. This field can only take the value as `yes-really-update-store`. If the user wants to change the `store.type` field for an existing cluster, then they will also need to update `spec.storage.store.updateStore` with `yes-really-update-store`
+
+
+- Add `status.ceph.osd` to the ceph cluster status. This will help convey the overall status of the OSD migration
+
+  ``` yaml
+  status:
+   ceph:
+      osd:
+       count: 8
```

Comment:
To keep the status simpler with only critical info, I'd suggest no need for the osd count. They can infer the count from the total of different types.

Reply:
removed `count`. Also keeping this API change under `status.storage.osd` rather than `status.ceph.osd`, like we discussed in yesterday's huddle.